drivers/sensor/ite_vcmp_it8xxx2/vcmp_ite_it8xxx2.c

```diff
@@ -0,0 +1,398 @@
+	void (*irq_cfg_func)(void);
+	/* Voltage comparator channel */
+	int vcmp_ch;
+	/* Scan period for "all voltage comparator channel" */
+	int scan_period;
+	/*
+	 * Determines the condition between ADC data and threshold_mv
+	 * that will trigger voltage comparator interrupt.
+	 */
+	int comparison;
+	/* Threshold assert value in mv */
+	int threshold_mv;
+	/* Pointer of ADC device that will be performing measurement */
+	const struct device *adc;
+};
+
+/* Driver data */
+struct vcmp_it8xxx2_data {
+	/* ADC channel config */
+	struct adc_channel_cfg adc_ch_cfg;
+	/* Work queue to be notified when threshold assertion happens */
+	struct k_work work;
+	/* Sensor trigger hanlder to notify user of assetion */
+	sensor_trigger_handler_t handler;
+	/* Pointer of voltage comparator device */
+	const struct device *vcmp;
+};
+
+/* Voltage comparator work queue address */
+static uint32_t vcmp_work_addr[VCMP_CHANNEL_CNT];
+
+static void clear_vcmp_status(const struct device *dev, int vcmp_ch)
+{
+	const struct vcmp_it8xxx2_config *const config = dev->config;
+	volatile uint8_t *reg_vcmpsts = config->reg_vcmpsts;
+	volatile uint8_t *reg_vcmpsts2 = config->reg_vcmpsts2;
+
+	/* W/C voltage comparator specific channel interrupt status */
+	if (vcmp_ch <= VCMP_CHANNEL_2) {
+		*reg_vcmpsts = BIT(vcmp_ch);
+	} else {
+		*reg_vcmpsts2 = BIT(vcmp_ch - VCMP_CHANNEL_3);
+	}
+}
+
+static void vcmp_enable(const struct device *dev, int enable)
+{
+	const struct vcmp_it8xxx2_config *const config = dev->config;
+	volatile uint8_t *reg_vcmpxctl = config->reg_vcmpxctl;
+
+	if (enable) {
+		/* Enable voltage comparator specific channel interrupt */
+		*reg_vcmpxctl |= IT8XXX2_VCMP_CMPINTEN;
+		/* Start voltage comparator specific channel */
+		*reg_vcmpxctl |= IT8XXX2_VCMP_CMPEN;
+	} else {
+		/* Stop voltage comparator specific channel */
+		*reg_vcmpxctl &= ~IT8XXX2_VCMP_CMPEN;
+		/* Disable voltage comparator specific channel interrupt */
+		*reg_vcmpxctl &= ~IT8XXX2_VCMP_CMPINTEN;
+	}
+}
+
+static void it8xxx2_vcmp_trigger_work_handler(struct k_work *item)
+{
+	struct vcmp_it8xxx2_data *data =
+			CONTAINER_OF(item, struct vcmp_it8xxx2_data, work);
+	struct sensor_trigger trigger = {
+		.type = SENSOR_TRIG_THRESHOLD,
+		.chan = SENSOR_CHAN_VOLTAGE
+	};
+
+	if (data->handler) {
+		data->handler(data->vcmp, &trigger);
+	}
+}
+
+static int vcmp_ite_it8xxx2_attr_set(const struct device *dev,
+				     enum sensor_channel chan,
+				     enum sensor_attribute attr,
+				     const struct sensor_value *val)
+{
+	const struct vcmp_it8xxx2_config *const config = dev->config;
+	volatile uint8_t *reg_vcmpxthrdatm = config->reg_vcmpxthrdatm;
+	volatile uint8_t *reg_vcmpxthrdatl = config->reg_vcmpxthrdatl;
+	volatile uint8_t *reg_vcmpxctl = config->reg_vcmpxctl;
+	int reg_val, ret = 0;
+
+	if (chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	switch ((uint16_t)attr) {
+	case SENSOR_ATTR_LOWER_THRESH:
+	case SENSOR_ATTR_UPPER_THRESH:
```

Comment:
`SENSOR_ATTR_LOWER/UPPER_THRESH` and `SENSOR_ATTR_LOWER/UPPER_VOLTAGE_THRESH` duplicate a lot of code.

This can be simplified as follows:

```
static int vcpm_ite_set_threshold(const struct device *dev, int reg_val)
{ ... }

static int vcmp_ite_it8xxx2_attr_set(...)
{
	switch ((uint16_t)attr) {
	case SENSOR_ATTR_LOWER_THRESH:
	case SENSOR_ATTR_UPPER_THRESH:
		ret = vcpm_ite_set_threshold(config, val->val1);
		break;
	case SENSOR_ATTR_LOWER_VOLTAGE_THRESH:
	case SENSOR_ATTR_UPPER_VOLTAGE_THRESH:
		reg_val = (val->val1 * VCMP_RESOLUTION / VCMP_MAX_MVOLT);
		ret = vcpm_ite_set_threshold(config, reg_val);
		break;
```

Alternately, you could eliminate `SENSOR_ATTR_LOWER/UPPER_VOLTAGE_THRESH` by doing the conversion from mv to the raw value inside `vcmp_it8xxx2_init()`.

Reply:
Simplified the duplicate code.
I'm not sure about eliminating SENSOR_ATTR_LOWER/UPPER_VOLTAGE_THRESH, does it never pass the "mv voltage" to vcmp_ite_it8xxx2_attr_set() in run-time?